datasets/pcfg/data_generation.py

```diff
@@ -0,0 +1,160 @@
+""" Generating training and test files for iterative decoding on PCFG.
+"""
+
+from absl import app
+import numpy as np
+import re
+
+# Specify local path to PCFG data in
+# https://github.com/i-machine-think/am-i-compositional/tree/master/data/pcfgset
+# Choose between iid (pcfgset), productivity and systematicity splits
+data_path = "/content/drive/MyDrive/preprocess_pcfg/pcfgset/data/"
+
+
+def generate_intermediate_tasks(source_line, target_line):
+  """Generates intermediate tasks for iterative decoding.
+
+  Generates a list of intermediate tasks (including original instruction 
+  and original output) for each source line and target line pair
+
+  Args:
+    source_line: A source (instruction) string from the PCFG dataset.
+    target_line: A target (output) string from the PCFG dataset.
+
+  Returns:
+    A list of strings where each string corresponds to an intermediate decoding 
+    task. The last string is always the final output + 'END'. For example:
+
+    ["swap_first_last copy remove_second E18 E15 Q6 , P15 L18 X10 I15 Y14",
+    "swap_first_last copy E18 E15 Q6", "swap_first_last E18 E15 Q6",
+    "Q6 E15 E18 END"]
+
+  Raises:
+    AssertionError: Output and target strings do not match.
+  """
+  
+  unary_keys = ['copy', 'reverse', 'shift', 'swap_first_last', 'repeat', 'echo']
+  binary_keys =['append', 'prepend', 'remove_first', 'remove_second']
```

Comment:
Nit: missing space in `=[`.
```suggestion
  binary_keys = ['append', 'prepend', 'remove_first', 'remove_second']
```